polaris-react/src/components/Button/Button.module.scss

```diff
@@ -377,7 +377,9 @@
 
 // BUTTON GROUP
 /* stylelint-disable -- selector-max-combinators */
-[data-buttongroup-variant='segmented'] > *:not(:first-child) .Button {
+[data-buttongroup-variant='segmented']
+  > *:not(:first-child)
+  .Button:is(.variantPrimary) {
```

Comment:
Having the negative left margin on all Buttons was causing styling issues for ButtonGroups with `connectedTop` where the ButtonGroup width was `1px` off of the sibling element that it was supposed to match border styles with.

![margin2](https://github.com/Shopify/polaris/assets/26749317/686d480e-465b-477c-b548-65593779b9eb)

I did some digging for context into when the negative margin was implemented and it looks like it was intended to [only apply](https://github.com/Shopify/polaris/blob/9484d55ca7a5b99f0cbb02676fd0f28fe92fc3f5/polaris-react/src/components/Button/Button.module.scss#L1017-L1020) to Buttons with `variant="primary"` to get rid of a visual double border caused by the inset. I modified the selector to include that check as it would solve the `1px` misalignment for `connectedTop` ButtonGroups.